File: libraries/python/src/urc_converter/sources/qua.py
"""Quaver (.qua) source parser and converter."""

from dataclasses import dataclass, field

import yaml

from ..error import UrcError
from ..model import Chart, Layout, Metadata, Note, NoteType, Version
from ._shared import build_timing, check_hold_overlap, round_ms

_MODE_KEYS = {1: 4, 2: 7, 3: 1, 4: 2, 5: 3, 6: 5, 7: 6, 8: 8, 9: 9, 10: 10}


@dataclass
class QuaTimingPoint:
	"""One TimingPoints entry."""

	start_time: int
	bpm: float
	signature: int


@dataclass
class QuaSvPoint:
	"""One ScrollSpeedFactors entry."""

	start_time: int
	multiplier: float


@dataclass
class QuaHitObject:
	"""One HitObjects entry."""

	start_time: int
	lane: int
	end_time: int = 0
	mine: bool = False


@dataclass
class QuaMap:
	"""Source model of a .qua chart."""

	mode: int = 1
	has_scratch_key: bool = False
	title: str | None = None
	artist: str | None = None
	creator: str | None = None
	difficulty_name: str | None = None
	timing_points: list[QuaTimingPoint] = field(default_factory=list)
	sv_points: list[QuaSvPoint] = field(default_factory=list)
	hit_objects: list[QuaHitObject] = field(default_factory=list)


def parse_qua(text: str) -> QuaMap:
	"""Parse .qua YAML text into a source model."""
	try:
		data = yaml.safe_load(text)
	except yaml.YAMLError as error:
		raise UrcError("syntax", _yaml_line(error), f"invalid YAML: {error}") from None

	if not isinstance(data, dict):
		raise UrcError("syntax", 1, ".qua must be a YAML mapping")

	qua = QuaMap(
		mode=_int(data, "Mode", 1),
		has_scratch_key=bool(data.get("HasScratchKey", False)),
		title=_str(data, "Title"),
		artist=_str(data, "Artist"),
		creator=_str(data, "Creator"),
		difficulty_name=_str(data, "DifficultyName"),
	)

	for entry in _entries(data, "TimingPoints"):
		signature = _int(entry, "Signature", 4)
		if signature == 0:  # legacy unset value, restored to 4/4 by the game
			signature = 4
		if signature not in (3, 4):
			raise UrcError("syntax", 1, f"unsupported time signature: {signature}")

		qua.timing_points.append(
			QuaTimingPoint(
				start_time=round_ms(_float(entry, "StartTime")),
				bpm=_float(entry, "Bpm"),
				signature=signature,
			)
		)

	for entry in _entries(data, "ScrollSpeedFactors"):
		qua.sv_points.append(
			QuaSvPoint(
				start_time=round_ms(_float(entry, "StartTime")),
				multiplier=_float(entry, "Multiplier"),
			)
		)

	for entry in _entries(data, "HitObjects"):
		qua.hit_objects.append(
			QuaHitObject(
				start_time=_int(entry, "StartTime", -1),
				lane=_int(entry, "Lane", 0),
				end_time=int(entry.get("EndTime") or 0),
				mine=_int(entry, "Type", 0) == 1,
			)
		)

	return qua


def convert_qua(qua: QuaMap) -> Chart:
	"""Map a Quaver chart onto a URC chart."""
	keys = _MODE_KEYS.get(qua.mode)
	if keys is None:
		raise UrcError("unsupported-version", 1, f"unsupported Quaver mode: {qua.mode}")

	special_keys = 1 if qua.has_scratch_key else 0

	first_note_time = min((obj.start_time for obj in qua.hit_objects), default=0)

	timing = build_timing(
		bpm_points=[(point.start_time, point.bpm, point.signature) for point in qua.timing_points],
		sv_points=[(point.start_time, point.multiplier) for point in qua.sv_points],
		first_note_time=first_note_time,
		source=".qua",
	)

	total = keys + special_keys
	notes: list[Note] = []

	for obj in qua.hit_objects:
		lane = obj.lane - 1
		if not 0 <= lane < total:
			raise UrcError("syntax", 1, f"lane out of range: {obj.lane}")

		if obj.end_time and obj.end_time < obj.start_time:
			raise UrcError(
				"syntax", 1, f"hold ends before it starts: {obj.end_time} < {obj.start_time}"
			)

		if obj.end_time:
			notes.append(
				Note(timestamp_ms=obj.start_time - first_note_time, lane=lane, type=NoteType.LS)
			)
			notes.append(
				Note(timestamp_ms=obj.end_time - first_note_time, lane=lane, type=NoteType.LE)
			)
		elif obj.mine:
			notes.append(
				Note(timestamp_ms=obj.start_time - first_note_time, lane=lane, type=NoteType.M)
			)
		else:
			notes.append(
				Note(timestamp_ms=obj.start_time - first_note_time, lane=lane, type=NoteType.N)
			)

	check_hold_overlap(notes)

	missing = [
		label
		for label, value in (
			("Title", qua.title),
			("Artist", qua.artist),
			("Creator", qua.creator),
			("DifficultyName", qua.difficulty_name),
		)
		if not value
	]
	if missing:
		raise UrcError("syntax", 1, f"missing metadata: {', '.join(missing)}")

	return Chart(
		format_version=Version(major=1, minor=1),
		metadata=Metadata(
			original="Quaver",
			title=qua.title,
			artist=qua.artist,
			creator=qua.creator,
			version=qua.difficulty_name,
		),
		judgment=None,
		layout=Layout(
			keys=keys,
			special_keys=special_keys,
			special_lanes=[keys] if qua.has_scratch_key else None,
		),
		timing=timing,
		notes=notes,
	)


def _yaml_line(error: yaml.YAMLError) -> int:
	mark = getattr(error, "problem_mark", None)
	return mark.line + 1 if mark else 1


def _entries(data: dict, key: str) -> list[dict]:
	value = data.get(key)
	if value is None:
		return []
	if not isinstance(value, list) or any(not isinstance(entry, dict) for entry in value):
		raise UrcError("syntax", 1, f"{key} must be a list of mappings")
	return value


def _str(data: dict, key: str) -> str | None:
	value = data.get(key)
	return str(value) if value is not None else None


def _int(data: dict, key: str, default: int) -> int:
	value = data.get(key)
	if value is None:
		return default
	try:
		return int(value)
	except TypeError, ValueError:
		raise UrcError("syntax", 1, f"invalid {key}: {value!r}") from None


def _float(entry: dict, key: str) -> float:
	value = entry.get(key)
	if value is None:
		raise UrcError("syntax", 1, f"missing {key}")
	try:
		return float(value)
	except TypeError, ValueError:
		raise UrcError("syntax", 1, f"invalid {key}: {value!r}") from None

File: libraries/dotnet/UrcConverter/Sources/Qua.fs
namespace UrcConverter.Sources

open System
open System.Globalization
open UrcConverter

module Qua =

    open System.IO
    open FsToolkit.ErrorHandling
    open YamlDotNet.RepresentationModel

    let private modeKeys =
        [
            (1, 4)
            (2, 7)
            (3, 1)
            (4, 2)
            (5, 3)
            (6, 5)
            (7, 6)
            (8, 8)
            (9, 9)
            (10, 10)
        ]

    type TimingPoint =
        {
            StartTime: int
            Bpm: float
            Signature: int
        }

    type SvPoint = { StartTime: int; Multiplier: float }

    type HitObject =
        {
            StartTime: int
            Lane: int
            EndTime: int
            Mine: bool
        }

    type QuaMap =
        {
            Mode: int
            HasScratchKey: bool
            Title: string option
            Artist: string option
            Creator: string option
            DifficultyName: string option
            TimingPoints: TimingPoint list
            SvPoints: SvPoint list
            HitObjects: HitObject list
        }

    let private tryFind (mapping: YamlMappingNode) (key: string) : YamlNode option =
        let mutable value = Unchecked.defaultof<YamlNode>

        if mapping.Children.TryGetValue(YamlScalarNode(key), &value) then
            Some value
        else
            None

    let private number (mapping: YamlMappingNode) (key: string) : Result<float, UrcError> =
        match tryFind mapping key with
        | Some(:? YamlScalarNode as scalar) ->
            match
                Double.TryParse(scalar.Value, NumberStyles.Float, CultureInfo.InvariantCulture)
            with
            | true, value -> Result.Ok value
            | false, _ -> Result.Error(UrcError.Syntax(1, $"invalid {key}: '{scalar.Value}'"))
        | Some _ -> Result.Error(UrcError.Syntax(1, $"invalid {key}"))
        | None -> Result.Error(UrcError.Syntax(1, $"missing {key}"))

    let private numberOr
        (mapping: YamlMappingNode)
        (key: string)
        (fallback: float)
        : Result<float, UrcError> =
        match tryFind mapping key with
        | None -> Result.Ok fallback
        | Some _ -> number mapping key

    let private entries
        (mapping: YamlMappingNode)
        (key: string)
        : Result<YamlMappingNode list, UrcError> =
        match tryFind mapping key with
        | None -> Result.Ok []
        | Some(:? YamlSequenceNode as sequence) ->
            sequence.Children
            |> Seq.toList
            |> List.map (fun node ->
                match node with
                | :? YamlMappingNode as entry -> Result.Ok entry
                | _ -> Result.Error(UrcError.Syntax(1, $"{key} must be a list of mappings")))
            |> Shared.sequence
        | Some _ -> Result.Error(UrcError.Syntax(1, $"{key} must be a list of mappings"))

    let private scalarText (mapping: YamlMappingNode) (key: string) : string option =
        match tryFind mapping key with
        | Some(:? YamlScalarNode as scalar) -> Some scalar.Value
        | _ -> None

    let parse (text: string) : Result<QuaMap, UrcError> =
        let stream = YamlStream()

        let loaded =
            try
                use reader = new StringReader(text)
                stream.Load(reader)
                Result.Ok()
            with error ->
                let line =
                    match error with
                    | :? YamlDotNet.Core.YamlException as yaml -> int yaml.Start.Line
                    | _ -> 1

                Result.Error(UrcError.Syntax(line, $"invalid YAML: {error.Message}"))

        match loaded with
        | Result.Error error -> Result.Error error
        | Result.Ok() ->
            match stream.Documents |> Seq.toList with
            | doc :: _ ->
                match doc.RootNode with
                | :? YamlMappingNode as root ->
                    result {
                        let! mode = numberOr root "Mode" 1.0 |> Result.map int

                        let hasScratchKey =
                            scalarText root "HasScratchKey"
                            |> Option.map (fun value -> value = "true")
                            |> Option.defaultValue false

                        let! timingPoints =
                            entries root "TimingPoints"
                            |> Result.bind (fun list ->
                                list
                                |> List.map (fun entry ->
                                    result {
                                        let! bpm = number entry "Bpm"

                                        let! startTime =
                                            number entry "StartTime" |> Result.map Shared.roundMs

                                        let! signature =
                                            numberOr entry "Signature" 4.0 |> Result.map int

                                        let signature = if signature = 0 then 4 else signature

                                        if signature <> 3 && signature <> 4 then
                                            return!
                                                Result.Error(
                                                    UrcError.Syntax(
                                                        1,
                                                        $"unsupported time signature: {signature}"
                                                    )
                                                )

                                        return
                                            {
                                                StartTime = startTime
                                                Bpm = bpm
                                                Signature = signature
                                            }
                                    })
                                |> Shared.sequence)

                        let! svPoints =
                            entries root "ScrollSpeedFactors"
                            |> Result.bind (fun list ->
                                list
                                |> List.map (fun entry ->
                                    result {
                                        let! startTime =
                                            number entry "StartTime" |> Result.map Shared.roundMs

                                        let! multiplier = number entry "Multiplier"

                                        return
                                            {
                                                StartTime = startTime
                                                Multiplier = multiplier
                                            }
                                    })
                                |> Shared.sequence)

                        let! hitObjects =
                            entries root "HitObjects"
                            |> Result.bind (fun list ->
                                list
                                |> List.map (fun entry ->
                                    result {
                                        let! startTime =
                                            number entry "StartTime" |> Result.map Shared.roundMs

                                        let! lane = number entry "Lane" |> Result.map int

                                        let! endTime =
                                            numberOr entry "EndTime" 0.0
                                            |> Result.map Shared.roundMs

                                        let! objectType =
                                            numberOr entry "Type" 0.0 |> Result.map int

                                        return
                                            {
                                                StartTime = startTime
                                                Lane = lane
                                                EndTime = endTime
                                                Mine = objectType = 1
                                            }
                                    })
                                |> Shared.sequence)

                        return
                            {
                                Mode = mode
                                HasScratchKey = hasScratchKey
                                Title = scalarText root "Title"
                                Artist = scalarText root "Artist"
                                Creator = scalarText root "Creator"
                                DifficultyName = scalarText root "DifficultyName"
                                TimingPoints = timingPoints
                                SvPoints = svPoints
                                HitObjects = hitObjects
                            }
                    }
                | _ -> Result.Error(UrcError.Syntax(1, ".qua must be a YAML mapping"))
            | [] -> Result.Error(UrcError.Syntax(1, ".qua has no YAML document"))

    let convert (qua: QuaMap) : Result<Chart, UrcError> =
        result {
            match modeKeys |> List.tryFind (fun (mode, _) -> mode = qua.Mode) with
            | None ->
                return!
                    Result.Error(
                        UrcError.UnsupportedVersion(1, $"unsupported Quaver mode: {qua.Mode}")
                    )
            | Some(_, keys) ->
                let specialKeys = if qua.HasScratchKey then 1 else 0

                let firstNoteTime =
                    match qua.HitObjects with
                    | [] -> 0
                    | objects -> objects |> List.map _.StartTime |> List.min

                let! timing =
                    Shared.buildTiming
                        (qua.TimingPoints
                         |> List.map (fun point -> point.StartTime, point.Bpm, point.Signature))
                        (qua.SvPoints |> List.map (fun point -> point.StartTime, point.Multiplier))
                        firstNoteTime
                        ".qua"

                let total = keys + specialKeys

                let! notes =
                    qua.HitObjects
                    |> List.map (fun obj ->
                        let lane = obj.Lane - 1

                        if lane < 0 || lane >= total then
                            Result.Error(UrcError.Syntax(1, $"lane out of range: {obj.Lane}"))
                        elif obj.EndTime <> 0 && obj.EndTime < obj.StartTime then
                            Result.Error(
                                UrcError.Syntax(
                                    1,
                                    $"hold ends before it starts: {obj.EndTime} < {obj.StartTime}"
                                )
                            )
                        elif obj.EndTime <> 0 then
                            Result.Ok
                                [
                                    {
                                        TimestampMs = obj.StartTime - firstNoteTime
                                        Lane = lane
                                        Type = NoteType.LS
                                    }
                                    {
                                        TimestampMs = obj.EndTime - firstNoteTime
                                        Lane = lane
                                        Type = NoteType.LE
                                    }
                                ]
                        elif obj.Mine then
                            Result.Ok
                                [
                                    {
                                        TimestampMs = obj.StartTime - firstNoteTime
                                        Lane = lane
                                        Type = NoteType.M
                                    }
                                ]
                        else
                            Result.Ok
                                [
                                    {
                                        TimestampMs = obj.StartTime - firstNoteTime
                                        Lane = lane
                                        Type = NoteType.N
                                    }
                                ])
                    |> Shared.sequence
                    |> Result.map List.concat

                do! Shared.checkHoldOverlap notes

                let metadata =
                    [
                        ("Title", qua.Title)
                        ("Artist", qua.Artist)
                        ("Creator", qua.Creator)
                        ("DifficultyName", qua.DifficultyName)
                    ]

                let missing =
                    metadata
                    |> List.choose (fun (name, value) ->
                        match value with
                        | Some value when value <> "" -> None
                        | _ -> Some name)

                let missingText = String.Join(", ", missing)

                if missing <> [] then
                    return! Result.Error(UrcError.Syntax(1, $"missing metadata: {missingText}"))

                match qua.Title, qua.Artist, qua.Creator, qua.DifficultyName with
                | Some title, Some artist, Some creator, Some version ->
                    return
                        {
                            FormatVersion = { Major = 1; Minor = 1 }
                            Metadata =
                                {
                                    Original = "Quaver"
                                    Title = title
                                    Artist = artist
                                    Creator = creator
                                    Version = version
                                }
                            Judgment = None
                            Layout =
                                {
                                    Keys = keys
                                    SpecialKeys = specialKeys
                                    SpecialLanes = if qua.HasScratchKey then Some [ keys ] else None
                                }
                            TimingPoints = timing
                            Notes = notes
                        }
                | _ -> return! Result.Error(UrcError.Syntax(1, $"missing metadata: {missingText}"))
        }
